### PR TITLE
[RLlib] Fix env seeding in EnvRunners and rewrite `deterministic_training` example script (was still on old stack).

### PR DESCRIPTION
**Pull request by sven1977**
* Fix env seeding in EnvRunners (properly seed env and module based on config.seed + worker_idx + some constant for eval runners).
* Rewrite `deterministic_training` example script (was still on old stack).

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/rllib/rllib-examples.rst [hunk 1]
```diff
@@ -170,6 +170,16 @@ Curriculum learning
...
 
 
+Debugging
++++++++++
+
+- `Deterministic sampling and training <https://github.com/ray-project/ray/blob/master/rllib/examples/debugging/deterministic_training.py>`__:
+   Demonstrates the possibility to seed an experiment through the algorithm config. RLlib passes the seed through to all components that have a copy of the
+   :ref:`RL environment <rllib-environments-doc>` and the :ref:`RLModule <rlmodule-guide>` and thus makes sure these components behave deterministically.
+   When using a seed, train results should become repeatable. Note that some algorithms, such as :ref:`APPO <appo>` which rely on asynchronous sampling
+   in combination with Ray network communication always behave stochastically, no matter whether you set a seed or not.
```
Comment (simonsays1980) [inline doc/source/rllib/rllib-examples.rst:180]: I think the experiences would still differ because of the action sampling via PyTorch, wouldn't they?

File: rllib/BUILD [hunk 2]
```diff
@@ -3954,19 +3954,35 @@ py_test(
...
+    tags = [
+        "examples",
+        "exclusive",
+        "multi_gpu",
+        "team:rllib",
+    ],
+)
+
+py_test(
+    name = "examples/debugging/deterministic_sampling_and_training_multi_agent",
```
Comment (simonsays1980) [inline rllib/BUILD:3976]: Awesome!

File: rllib/env/env_runner.py [hunk 3]
```diff
@@ -69,6 +75,22 @@ def __init__(self, *, config: "AlgorithmConfig", **kwargs):
         ):
             tf1.enable_eager_execution()
 
+        # Determine actual seed for this particular worker based on worker index AND
+        # whether it's an eval worker.
+        self._seed: Optional[int] = None
+        if self.config.seed is not None:
+            self._seed = int(
+                self.config.seed
+                + (self.worker_index or 0)
```
Comment (simonsays1980) [inline rllib/env/env_runner.py:86]: Dumb question: The worker index stays the same when a work crashes and is replaced, correct?

File: rllib/utils/debug/deterministic.py [hunk 4]
```diff
@@ -32,20 +33,27 @@ def update_global_seed_if_necessary(
...
             os.environ["CUBLAS_WORKSPACE_CONFIG"] = "4096:8"
+            torch.cuda.manual_seed(seed)
+            torch.cuda.manual_seed_all(seed)  # if using multi-GPU
+
+        if Version(torch.__version__) >= Version("1.8.0"):
+            # Not all Operations support this.
+            torch.use_deterministic_algorithms(True)
         else:
-            from packaging.version import Version
+            torch.set_deterministic(True)
```
Comment (simonsays1980) [inline rllib/utils/debug/deterministic.py]: Great!

**Review by simonsays1980 — APPROVED** [review]
LGTM. Just small nits.

File: rllib/env/env_runner.py (hunk 3 above)
Comment (sven1977) [reply rllib/env/env_runner.py:86]: Correct. They never change.

File: doc/source/rllib/rllib-examples.rst (hunk 1 above)
Comment (sven1977) [reply doc/source/rllib/rllib-examples.rst:180]: But the torch is also seeded, no?
The new example script test proves that it's ok.
Runs on 2 GPUs in the CI, so this should cover all weird cases.
